File: python/src/models/guardians.py
from abc import ABCMeta, abstractmethod
from enum import Enum
from .mixins import *


class GuardianPose(Enum):
  STAND     = "stand"
  CORRECT   = "correct"
  INCORRECT = "incorrect"


class Guardian():
  """Parent Guardian Class. All Guardians derive from this class."""

  def __init__(self, name, description):
    self.name           = name
    self.description    = description
    self.shrine         = None
    self.pose           = GuardianPose.STAND

    self.try_count      = 3
    self.question_count = 5
    self.question       = None
    self.answer         = None
    self.required_item  = "Calculator"

  def prep_quiz(self, player):
    """
    Preps for the battle.

    If the player runs away and returns,
    this will make sure to reset the tries, questions, and required item.
    """
    self.player = player
    self.try_count = 3
    self.question_count = 5
    self.question = None
    self.required_item  = "Calculator"

  def create_question(self):
    """
    Resets the guardian pose, and creates a new question.

    If there is already a question, does nothing.
    If there is a required item, asks for item.
    All else, creates and asks a new math question.
    """
    self.pose = GuardianPose.STAND
    if self.question is not None:
      return

    if self.required_item is not None:
      self.question = f"{self.name} is requesting you use ({self.required_item})"
    else:
      self.math_question()

  def math_question(self):
    raise NotImplementedError

  def next_question_prep(self):
    """
    Preps for the next question.

    This includes decreasing the question_count,
    and setting a required_item if necessary.
    """
    raise NotImplementedError

  def check_item(self, name):
    """
    Checks if the used item matches what is required.

    If:
    - Item matches required item   → True
    - Item does not match required → False
    """
    if name == self.required_item:
      self.required_item = None
      self.question = None
      self.pose = GuardianPose.CORRECT
      return True
    else:
      self.pose = GuardianPose.INCORRECT
      self.try_count -= 1
      return False

  def check_answer(self, answer):
    """
    Checks if the answer matches self.answer.

    If the answer is correct,
    resets the answer,
    calls next_question_prep(),
    and returns true,

    Else, returns false.
    """
    if answer == self.answer:
      self.answer = None
      self.question = None
      self.pose = GuardianPose.CORRECT
      self.next_question_prep()
      return True
    else:
      self.pose = GuardianPose.INCORRECT
      self.try_count -= 1
      return False

  def check_victory(self):
    """
    Checks if the player has won/lost the battle.

    If the player loses:
    return False

    If the player wins:
    detach the guardian from the Shrine and return True

    Still battling? Return None
    """
    if self.try_count == 0:
      return False
    elif self.question_count == 0:
      self.shrine.guardian = None
      return True
    else:
      return None


class MultipGuardian(CanAdd, Guardian):
  """Asks addition questions. Drops the Multip sign."""

  def __init__(self):
    super().__init__("Multip Guardian",
                     "Guardian of the Multip Shrine. Keep your guard!")

  def math_question(self):
    """
    Creates a unique question for each question_count:
    - 5 → addition(easy)
    - 4 → addition(easy)
    - 3 → addition(med)
    - 2 → addition(hard)
    - 1 → addition(hard)
    """
    if self.question_count > 3:
      # First two questions? Difficulty easy
      problem = self.addition(0)
    elif self.question_count > 2:
      # Third question? Difficulty medium
      problem = self.addition(1)
    else:
      # Last two questions? Difficulty hard
      problem = self.addition(2)
    [self.question, self.answer] = problem

  def next_question_prep(self):
    self.question_count -= 1


class DividGuardian(CanSubtract, Guardian):

  def __init__(self):
    super().__init__("Divid Guardian",
                     "Guardian of the Divid Shrine. Keep it sleek, slick.")

  def math_question(self):
    """
    Creates a unique question for each question_count:
    - 5 → subtraction(easy)
    - 4 → subtraction(easy)
    - 3 → subtraction(med)
    - 2 → subtraction(hard)
    - 1 → subtraction(hard)
    """
    if self.question_count > 3:
      # First two questions? Difficulty easy
      problem = self.subtraction(0)
    elif self.question_count > 2:
      # Third question? Difficulty medium
      problem = self.subtraction(1)
    else:
      # Last two questions? Difficulty hard
      problem = self.subtraction(2)
    [self.question, self.answer] = problem

  def next_question_prep(self):
    self.question_count -= 1


class SquareGuardian(CanAdd, CanMultiply, Guardian):

  def __init__(self):
    super().__init__("Square Guardian",
                     "Guardian of the Square Shrine. Stay happy, stay civil!")

  def math_question(self):
    """
    Creates a unique question for each question_count:
    - 5 → addition(easy)
    - 4 → addition(med)
    - 3 → addition(hard)
    - 2 → multiplication(easy)
    - 1 → multiplication(hard)
    """
    if self.question_count == 5:
      problem = self.addition(0)
    elif self.question_count == 4:
      problem = self.addition(1)
    elif self.question_count == 3:
      problem = self.addition(2)
    elif self.question_count == 2:
      problem = self.multiplication(0)
    else:
      problem = self.multiplication(1)

    [self.question, self.answer] = problem

  def next_question_prep(self):
    self.question_count -= 1
    if self.question_count == 2:
      self.required_item = "Multip"


class RadicalGuardian(CanAdd, CanSubtract, CanSquare, CanDivide, Guardian):

  def __init__(self):
    super().__init__("Radical Guardian",
                     "Guardian of the Radical Shrine. Radical!!!")

  def math_question(self):
    """
    Creates a unique question for each question_count:
    - 5 → addition(med)
    - 4 → subtraction(hard)
    - 3 → square(hard)
    - 2 → divide(med)
    - 1 → divide(hard)
    """
    if self.question_count == 5:
      problem = self.addition(1)
    elif self.question_count == 4:
      problem = self.subtraction(2)
    elif self.question_count == 3:
      problem = self.square(2)
    elif self.question_count == 2:
      problem = self.division(1)
    else:
      problem = self.division(2)

    [self.question, self.answer] = problem

  def next_question_prep(self):
    self.question_count -= 1
    if self.question_count == 3:
      self.required_item = "Square"

    if self.question_count == 2:
      self.required_item = "Divid"


class ArtifactGuardian(CanMultiply, CanSquare, CanDivide, CanRoot, CanAdd, CanSubtract, Guardian):

  def __init__(self):
    super().__init__("Artifact Guardian",
                     "The final challenge! If you run away, your questions won't get reset.")
    # The artifact Guardian holds it's own questions and tries.
    # You can run away and the question count won't go down.
    self.question_count = 10
    self.try_count      = 3
    self.required_item  = "Multip"

  def prep_quiz(self, player):
    """
    Sets the battle's try count and question count
    to the Guardian's saved values.

    This allows the player to run away and come back
    without resetting their questions. 
    """
    self.player          = player
    self.question        = None
    self.try_count       = 3
    self.question_count += 1
    # We set question_count + 1 because next_question_prep
    # will decrease it by one automatically
    self.next_question_prep()

  def math_question(self):
    """
    Creates a unique question for each question_count:
    - 10 → multiplication(hard)
    - 9  → multiplication(hard)
    - 8  → square(med)
    - 7  → square(hard)
    - 6  → division(hard)
    - 5  → division(hard)
    - 4  → root(med)
    - 3  → root(hard)
    - 2  → addition(hard)
    - 1  → subtraction(hard)
    """
    if self.question_count > 8:
      problem = self.multiplication(2)

    elif self.question_count > 7:
      problem = self.square(1)
    elif self.question_count > 6:
      problem = self.square(2)

    elif self.question_count > 4:
      problem = self.division(2)

    elif self.question_count > 3:
      problem = self.root(1)
    elif self.question_count > 2:
      problem = self.root(2)

    elif self.question_count > 1:
      problem = self.addition(2)
    else:
      problem = self.subtraction(2)

    [self.question, self.answer] = problem

  def next_question_prep(self):
    self.question_count -= 1
    if self.question_count == 10:
      self.required_item = "Multip"

    if self.question_count == 8:
      self.required_item = "Square"

    if self.question_count == 6:
      self.required_item = "Divid"

    if self.question_count == 4:
      self.required_item = "Radical"

    if self.question_count == 2:
      self.required_item = "Calculator"

File: JS/src/models/Guardians.ts
import {
  buildAdditionQuestion,
  buildSubtractionQuestion,
  buildDivisionQuestion,
  buildMultiplicationQuestion,
  buildExponentQuestion,
  buildRootQuestion,
} from "./guardianUtils";

export enum GuardianPose {
  STAND = "stand",
  CORRECT = "correct",
  INCORRECT = "incorrect",
}

export enum GuardianName {
  MULTIP = "Multip Guardian",
  DIVID = "Divid Guardian",
  SQUARE = "Square Guardian",
  RADICAL = "Radical Guardian",
  ARTIFACT = "Artifact Guardian",
}

export interface GuardianQuestion {
  mathQuestion: string;
  mathAnswer: number;
  itemPrerequisite?: string;
}

/**Parent Guardian Class. All Guardians derive from this class. */
export abstract class Guardian {
  constructor(
    readonly name: GuardianName,
    readonly description: string,

    private __triesLeft: number = 3,
    private __isAlive: boolean = true,
    protected guardianQuestions: GuardianQuestion[] = [],
  ) {}

  get triesLeft() {
    return this.__triesLeft;
  }

  public decrementTriesLeft() {
    this.__triesLeft--;
  }

  get questionsRemaining() {
    return this.guardianQuestions.length;
  }

  /** Removes the current guardian question, also decrementing the total number of questions */
  public removeCurrentQuestion() {
    if (this.guardianQuestions[0].itemPrerequisite) {
      this.guardianQuestions[0].itemPrerequisite = undefined;
    } else {
      this.guardianQuestions.shift();
    }
  }

  /**
   * Returns the expected player response
   *
   * This could be a number or a string depending on
   * whether the player needs to use an item or answer a math question.
   */
  get correctAnswer(): string | number {
    const currentQuestion = this.guardianQuestions[0];

    return currentQuestion.itemPrerequisite?.toLowerCase() ?? currentQuestion.mathAnswer;
  }

  get isAlive() {
    return this.__isAlive;
  }

  public die() {
    this.__isAlive = false;
  }

  /**
   * Returns the question to display to the player.
   *
   * If there's no question, we'll create one here.
   *
   * If there's a required item, that becomes our question.
   * Otherwise, our question is a math question
   */
  public getNextQuestion() {
    if (this.guardianQuestions[0].itemPrerequisite) {
      return `~W${this.name}~e is requesting you use ~c(${this.guardianQuestions[0].itemPrerequisite})`;
    } else return this.guardianQuestions[0].mathQuestion;
  }

  /**
   * Resets the user's tries.
   *
   * This function SHOULD also set the guardianQuestions.
   */
  public prepareQuestions() {
    this.__triesLeft = 3;
  }
}

/**Asks addition questions. Drops the Multip sign.*/
export class MultipGuardian extends Guardian {
  constructor() {
    super(GuardianName.MULTIP, "Guardian of the Multip Shrine. Keep your guard!");
  }

  /**
   * Creates questions in this order:
   * - 1st → addition(easy), requires calculator
   * - 2nd → addition(easy)
   * - 3rd → addition(med)
   * - 4th → addition(hard)
   * - 5th → addition(hard)
   */
  public prepareQuestions(): void {
    super.prepareQuestions();
    this.guardianQuestions = [
      { ...buildAdditionQuestion(0), itemPrerequisite: "Calculator" },
      buildAdditionQuestion(0),
      buildAdditionQuestion(1),
      buildAdditionQuestion(2),
      buildAdditionQuestion(2),
    ];
  }
}

export class DividGuardian extends Guardian {
  constructor() {
    super(GuardianName.DIVID, "Guardian of the Divid Shrine. Keep it sleek, slick.");
  }

  /**
   * Creates questions in this order:
   * - 1st → subtraction(easy), requires calculator
   * - 2nd → subtraction(easy)
   * - 3rd → subtraction(med)
   * - 4th → subtraction(hard)
   * - 5th → subtraction(hard)
   */
  public prepareQuestions(): void {
    super.prepareQuestions();
    this.guardianQuestions = [
      { ...buildSubtractionQuestion(0), itemPrerequisite: "Calculator" },
      buildSubtractionQuestion(0),
      buildSubtractionQuestion(1),
      buildSubtractionQuestion(2),
      buildSubtractionQuestion(2),
    ];
  }
}

export class SquareGuardian extends Guardian {
  constructor() {
    super(GuardianName.SQUARE, "Guardian of the Square Shrine. Stay happy, stay civil!");
  }

  /**
   * Creates questions in this order:
   * - 1st → addition(easy), requires calculator
   * - 2nd → addition(med)
   * - 3rd → addition(hard)
   * - 4th → multiplication(easy), requires Multip
   * - 5th → multiplication(hard)
   */
  public prepareQuestions(): void {
    super.prepareQuestions();
    this.guardianQuestions = [
      { ...buildAdditionQuestion(0), itemPrerequisite: "Calculator" },
      buildAdditionQuestion(1),
      buildAdditionQuestion(2),
      { ...buildMultiplicationQuestion(0), itemPrerequisite: "Multip" },
      buildMultiplicationQuestion(2),
    ];
  }
}

export class RadicalGuardian extends Guardian {
  constructor() {
    super(GuardianName.RADICAL, "Guardian of the Radical Shrine. Radical!!!");
  }

  /**
   * Creates questions in this order:
   * - 1st → addition(med), requires calculator
   * - 2nd → subtraction(hard)
   * - 3rd → square(hard), requires Square
   * - 4th → divide(med), requires Divid
   * - 5th → divide(hard)
   */
  public prepareQuestions(): void {
    super.prepareQuestions();
    this.guardianQuestions = [
      { ...buildAdditionQuestion(1), itemPrerequisite: "Calculator" },
      buildSubtractionQuestion(2),
      { ...buildExponentQuestion(2), itemPrerequisite: "Square" },
      { ...buildDivisionQuestion(1), itemPrerequisite: "Divid" },
      buildDivisionQuestion(2),
    ];
  }
}

/**
 * The artifact Guardian holds it's own questions and tries.
 * You can run away and the question count won't go down.
 */
export class ArtifactGuardian extends Guardian {
  constructor() {
    super(GuardianName.ARTIFACT, "The final challenge! If you run away, your questions won't get reset.");
  }

  /**
   * Creates questions in this order:
   * - 1st  → multiplication(hard), requires Multip
   * - 2nd  → multiplication(hard)
   * - 3rd  → square(med), requires Square
   * - 4th  → square(hard)
   * - 5th  → division(hard), requires Divid
   * - 6th  → division(hard)
   * - 7th  → root(med), requires Radical
   * - 8th  → root(hard)
   * - 9th  → addition(hard), requires calculator
   * - 10th → subtraction(hard)
   */
  public prepareQuestions(): void {
    super.prepareQuestions();
    if (this.guardianQuestions.length !== 0) {
      // The artifact guardian doesn't reset its questions.
      // Ergo, if there are already questions on the guardian then we don't do anything.
      return;
    }
    this.guardianQuestions = [
      { ...buildMultiplicationQuestion(2), itemPrerequisite: "Multip" },
      buildMultiplicationQuestion(2),
      { ...buildExponentQuestion(1), itemPrerequisite: "Square" },
      buildExponentQuestion(2),
      { ...buildDivisionQuestion(2), itemPrerequisite: "Divid" },
      buildDivisionQuestion(2),
      { ...buildRootQuestion(1), itemPrerequisite: "Radical" },
      buildRootQuestion(2),
      { ...buildAdditionQuestion(2), itemPrerequisite: "Calculator" },
      buildSubtractionQuestion(2),
    ];
  }
}
